File: src/wrapper.js
import { Session } from "./session.js";
import { normalizeResponse } from "./normalizer.js";
import { rewriteUrls } from "./tracking.js";
import {
  hostedSearch,
  injectMemories,
} from "../packages/memory/src/hosted.js";

// Default memory-injection knobs. Match the proxy's defaults so SDK and
// proxy customers see identical retrieval behaviour.
const MEMORY_DEFAULTS = {
  limit: 6,
  minScore: 0.55,
  timeoutMs: 800,
};

/**
 * Detect the client type by duck-typing its shape.
 */
function detectClientType(client) {
  if (client?.chat?.completions?.create) return "openai";
  if (client?.messages?.create) return "anthropic";
  if (typeof client?.run === "function") return "workers-ai";
  return "unknown";
}

/**
 * Pull the last user message from a request body. Anthropic + OpenAI both
 * carry messages on `params.messages`; Workers AI may also use
 * `params.prompt` or `params.input_text`. Returns null when nothing usable
 * is present (e.g. embedding call, empty prompt) so memory retrieval is
 * skipped cleanly.
 */
function extractLastUserMessage(params, provider) {
  // Only messages-shaped requests are eligible for system-prompt injection.
  // Workers AI prompt-style calls (`{ prompt: "..." }`) are passed through
  // unchanged — there's no clean place to insert memory context without
  // changing the request shape, and we never want to surprise the caller
  // by mutating their prompt string.
  void provider;
  const msgs = Array.isArray(params?.messages) ? params.messages : null;
  if (!msgs) return null;
  for (let i = msgs.length - 1; i >= 0; i--) {
    if (msgs[i].role === "user") {
      const c = msgs[i].content;
      if (typeof c === "string") return c;
      if (Array.isArray(c)) {
        return c
          .filter((p) => p.type === "text" && typeof p.text === "string")
          .map((p) => p.text)
          .join("\n");
      }
    }
  }
  return null;
}

/**
 * Inject memories from TES into request params before the LLM call.
 *
 * Default-on. Disable per-wrapClient via `sessionOpts.memory: false` or
 * per-call via `sessionOpts.memoryOpts.disable: true`. Knobs come from
 * `sessionOpts.memoryOpts` (`limit`, `minScore`, `timeoutMs`).
 *
 * Failure modes (TES timeout, module disabled, network error) are
 * non-fatal — the call proceeds with the customer's original params and
 * the skip reason is recorded on the session under `_lastMemoryStats`
 * for observability.
 */
async function maybeInjectMemories(
  clientConfig,
  sessionOpts,
  params,
  provider
) {
  if (sessionOpts.memory === false) {
    return { params, injected: 0, skipped: "memory_disabled" };
  }

  if (!clientConfig?.endpoint || !clientConfig?.apiKey) {
    return { params, injected: 0, skipped: "no_tes_config" };
  }

  const userMessage = extractLastUserMessage(params, provider);
  if (!userMessage) {
    return { params, injected: 0, skipped: "no_user_message" };
  }

  const opts = { ...MEMORY_DEFAULTS, ...(sessionOpts.memoryOpts || {}) };
  const { memories, skipped } = await hostedSearch(
    {
      endpoint: clientConfig.endpoint,
      clientId: clientConfig.clientId,
      apiKey: clientConfig.apiKey,
    },
    userMessage,
    opts
  );

  if (!memories?.length) {
    return { params, injected: 0, skipped: skipped || "no_memories" };
  }

  return {
    params: injectMemories(params, memories, provider),
    injected: memories.length,
    skipped: null,
  };
}

function recordMemoryStats(sessionOpts, stats) {
  if (sessionOpts._session) {
    sessionOpts._session._lastMemoryStats = stats;
  }
}

/**
 * Wrap any supported LLM client with automatic usage tracking.
 * Auto-detects OpenAI, Anthropic, and Workers AI clients.
 */
export function wrapClient(clientConfig, client, sessionOpts = {}) {
  // Resolve sessionId once so all calls share the same session
  sessionOpts._resolvedSessionId =
    sessionOpts.sessionId || crypto.randomUUID();

  // Shared session accumulates usage and tool calls across rounds
  sessionOpts._session = new Session(clientConfig, {
    sessionId: sessionOpts._resolvedSessionId,
    metadata: sessionOpts.metadata,
  });

  const type = detectClientType(client);

  if (type === "openai") return wrapOpenAI(clientConfig, client, sessionOpts);
  if (type === "anthropic")
    return wrapAnthropic(clientConfig, client, sessionOpts);
  if (type === "workers-ai")
    return wrapWorkersAI(clientConfig, client, sessionOpts);

  throw new Error(
    "Unsupported client: expected OpenAI (chat.completions.create), " +
      "Anthropic (messages.create), or Workers AI (run) client"
  );
}

// --- OpenAI ---

function wrapOpenAI(clientConfig, client, sessionOpts) {
  return new Proxy(client, {
    get(target, prop) {
      if (prop === "chat")
        return wrapOpenAIChat(clientConfig, target.chat, target, sessionOpts);
      if (prop === "sessionId") return sessionOpts._resolvedSessionId;
      if (prop === "tesSession") return sessionOpts._session;
      if (prop === "session")
        return (opts) => new OpenAISession(clientConfig, target, opts);
      return target[prop];
    },
  });
}

function wrapOpenAIChat(clientConfig, chat, client, sessionOpts) {
  return new Proxy(chat, {
    get(target, prop) {
      if (prop === "completions")
        return wrapOpenAICompletions(
          clientConfig,
          target.completions,
          client,
          sessionOpts
        );
      return target[prop];
    },
  });
}

function wrapOpenAICompletions(clientConfig, completions, client, sessionOpts) {
  return new Proxy(completions, {
    get(target, prop) {
      if (prop === "create") {
        return async (params) => {
          const memStats = await maybeInjectMemories(
            clientConfig,
            sessionOpts,
            params,
            "openai"
          );
          recordMemoryStats(sessionOpts, memStats);
          const result = await target.create(memStats.params);
          const content = result.choices?.[0]?.message?.content;
          if (content) {
            result.choices[0].message.content = await rewriteUrls(
              content,
              clientConfig,
              sessionOpts._resolvedSessionId,
              sessionOpts.metadata
            );
          }
          fireAndForgetEmit(
            clientConfig,
            sessionOpts,
            memStats.params.messages,
            result
          );
          return result;
        };
      }
      return target[prop];
    },
  });
}

class OpenAISession extends Session {
  constructor(clientConfig, client, opts) {
    super(clientConfig, opts);
    this._client = client;
  }

  async chat(params) {
    const result = await this._client.chat.completions.create(params);
    this.record(result);
    return result;
  }
}

// --- Anthropic ---

function wrapAnthropic(clientConfig, client, sessionOpts) {
  return new Proxy(client, {
    get(target, prop) {
      if (prop === "messages")
        return wrapAnthropicMessages(
          clientConfig,
          target.messages,
          target,
          sessionOpts
        );
      if (prop === "sessionId") return sessionOpts._resolvedSessionId;
      if (prop === "tesSession") return sessionOpts._session;
      if (prop === "session")
        return (opts) => new AnthropicSession(clientConfig, target, opts);
      return target[prop];
    },
  });
}

function wrapAnthropicMessages(clientConfig, messages, client, sessionOpts) {
  return new Proxy(messages, {
    get(target, prop) {
      if (prop === "create") {
        return async (params) => {
          const memStats = await maybeInjectMemories(
            clientConfig,
            sessionOpts,
            params,
            "anthropic"
          );
          recordMemoryStats(sessionOpts, memStats);
          const result = await target.create(memStats.params);
          if (Array.isArray(result.content)) {
            for (const block of result.content) {
              if (block.type === "text" && block.text) {
                block.text = await rewriteUrls(
                  block.text,
                  clientConfig,
                  sessionOpts._resolvedSessionId,
                  sessionOpts.metadata
                );
              }
            }
          }
          fireAndForgetEmit(
            clientConfig,
            sessionOpts,
            memStats.params.messages,
            result
          );
          return result;
        };
      }
      return target[prop];
    },
  });
}

class AnthropicSession extends Session {
  constructor(clientConfig, client, opts) {
    super(clientConfig, opts);
    this._client = client;
  }

  async chat(params) {
    const result = await this._client.messages.create(params);
    this.record(result);
    return result;
  }
}

// --- Workers AI ---

function wrapWorkersAI(clientConfig, aiBinding, sessionOpts) {
  return new Proxy(aiBinding, {
    get(target, prop) {
      if (prop === "run") {
        return async (model, params, ...rest) => {
          const memStats = await maybeInjectMemories(
            clientConfig,
            sessionOpts,
            params,
            "workers-ai"
          );
          recordMemoryStats(sessionOpts, memStats);
          const result = await target.run(model, memStats.params, ...rest);
          if (result.response) {
            result.response = await rewriteUrls(
              result.response,
              clientConfig,
              sessionOpts._resolvedSessionId,
              sessionOpts.metadata
            );
          }
          fireAndForgetEmit(
            clientConfig,
            sessionOpts,
            memStats.params?.messages,
            result,
            model
          );
          return result;
        };
      }
      if (prop === "sessionId") return sessionOpts._resolvedSessionId;
      if (prop === "tesSession") return sessionOpts._session;
      if (prop === "session")
        return (opts) => new WorkersAISession(clientConfig, target, opts);
      return target[prop];
    },
  });
}

class WorkersAISession extends Session {
  constructor(clientConfig, aiBinding, opts) {
    super(clientConfig, opts);
    this._ai = aiBinding;
  }

  async chat(model, params, ...rest) {
    const result = await this._ai.run(model, params, ...rest);
    this.record(result);
    return result;
  }
}

// --- Shared ---

/**
 * Extract tool results from the messages array and attach them to recorded
 * tool calls in the session. Messages contain {role:"tool", content, tool_call_id}
 * entries after the app executes tools and feeds results back to the AI.
 */
function extractToolResults(session, messages) {
  if (!messages?.length || !session._toolCalls.length) return;

  // Build map: tool_call_id -> tool name from assistant messages
  const idToName = new Map();
  for (const msg of messages) {
    if (msg.role === "assistant" && msg.tool_calls) {
      for (const tc of msg.tool_calls) {
        const id = tc.id || tc.tool_call_id;
        const name = tc.function?.name || tc.name;
        if (id && name) idToName.set(id, name);
      }
    }
  }

  // Attach results to session tool calls
  for (const msg of messages) {
    if (msg.role !== "tool" || !msg.content) continue;

    const callId = msg.tool_call_id;
    const toolName = callId ? idToName.get(callId) : null;

    // Find matching tool call in session (by name, without a result yet)
    for (const tc of session._toolCalls) {
      if (tc.result) continue; // already has a result
      if (toolName && tc.tool !== toolName) continue;

      // Parse JSON content if possible, otherwise store as string
      try {
        const parsed = JSON.parse(msg.content);
        // Summarise arrays (e.g. product lists) to avoid bloating the event
        if (Array.isArray(parsed)) {
          tc.result = { count: parsed.length, sample: parsed.slice(0, 3) };
        } else {
          tc.result = parsed;
        }
      } catch {
        tc.result = msg.content;
      }
      break;
    }
  }
}

function fireAndForgetEmit(clientConfig, sessionOpts, messages, result, model) {
  const session = sessionOpts._session;
  const normalized = session.record(result);

  // Extract tool results from the messages array.
  extractToolResults(session, messages);

  // Capture system prompt from messages (first system message, only once)
  if (!session._systemPrompt && messages?.length) {
    const systemMsg = messages.find((m) => m.role === "system");
    if (systemMsg?.content) {
      session._systemPrompt = systemMsg.content;
    }
  }

  // If Workers AI didn't include model in the response, use the one passed to run()
  if (model && !normalized.model) {
    session._model = model;
  }

  // When autoEmit is disabled, the caller controls event emission.
  // The wrapper still tracks usage/tool calls via session.record() above.
  if (sessionOpts.autoEmit === false) {
    return;
  }

  // Accumulate tool-call rounds without emitting — only emit when there's
  // actual text content (the final response in a multi-round tool loop).
  // This ensures a single CHAT_TURN event per conversation turn with all
  // accumulated tool calls and usage included.
  if (!normalized.content && normalized.toolCalls.length > 0) {
    return;
  }

  const userMsg =
    messages?.filter?.((m) => m.role === "user")?.pop()?.content || "";
  const assistantMsg = normalized.content || "";

  const emitPromise = session
    .emitChatTurn({ userMessage: userMsg, assistantResponse: assistantMsg })
    .catch((err) => console.error("[pentatonic-ai] emit failed:", err.message));

  // On Cloudflare Workers, the runtime terminates after the response is sent.
  // waitUntil keeps the Worker alive for background work like event emission.
  if (typeof sessionOpts.waitUntil === "function") {
    sessionOpts.waitUntil(emitPromise);
  }
}

File: python/pentatonic_agent_events/wrapper.py
import json
import sys
import uuid
from .normalizer import normalize_response
from .session import Session
from .tracking import rewrite_urls
from .transport import send_event


def detect_client_type(client):
    """Detect client type by duck-typing its shape."""
    try:
        if hasattr(client, "chat") and hasattr(client.chat, "completions") and hasattr(client.chat.completions, "create"):
            return "openai"
    except Exception:
        pass
    try:
        if hasattr(client, "messages") and hasattr(client.messages, "create"):
            return "anthropic"
    except Exception:
        pass
    if callable(getattr(client, "run", None)):
        return "workers-ai"
    return "unknown"


def wrap_client(config, client, session_id=None, metadata=None, auto_emit=True, wait_until=None):
    client_type = detect_client_type(client)
    sid = session_id or str(uuid.uuid4())
    meta = metadata or {}

    # Shared session accumulates usage and tool calls across rounds
    session = Session(config, session_id=sid, metadata=meta)

    opts = {
        "session": session,
        "auto_emit": auto_emit,
        "metadata": meta,
        "wait_until": wait_until,
    }

    if client_type == "openai":
        return _OpenAIWrapper(config, client, opts)
    if client_type == "anthropic":
        return _AnthropicWrapper(config, client, opts)
    if client_type == "workers-ai":
        return _WorkersAIWrapper(config, client, opts)

    raise ValueError(
        "Unsupported client: expected OpenAI (chat.completions.create), "
        "Anthropic (messages.create), or Workers AI (run) client"
    )


def _truncate(value, max_len):
    if not value or not max_len or not isinstance(value, str):
        return value
    if len(value) <= max_len:
        return value
    return value[:max_len] + "...[truncated]"


def _extract_tool_results(session, messages):
    """
    Extract tool results from the messages array and attach them to recorded
    tool calls in the session. Messages contain {role:"tool", content, tool_call_id}
    entries after the app executes tools and feeds results back to the AI.
    """
    if not messages or not session._tool_calls:
        return

    # Build map: tool_call_id -> tool name from assistant messages
    id_to_name = {}
    for msg in messages:
        if not isinstance(msg, dict):
            continue
        if msg.get("role") == "assistant" and msg.get("tool_calls"):
            for tc in msg["tool_calls"]:
                tc_id = tc.get("id") or tc.get("tool_call_id")
                name = None
                if isinstance(tc.get("function"), dict):
                    name = tc["function"].get("name")
                if not name:
                    name = tc.get("name")
                if tc_id and name:
                    id_to_name[tc_id] = name

    # Attach results to session tool calls
    for msg in messages:
        if not isinstance(msg, dict):
            continue
        if msg.get("role") != "tool" or not msg.get("content"):
            continue

        call_id = msg.get("tool_call_id")
        tool_name = id_to_name.get(call_id) if call_id else None

        # Find matching tool call in session (by name, without a result yet)
        for tc in session._tool_calls:
            if tc.get("result"):
                continue
            if tool_name and tc.get("tool") != tool_name:
                continue

            # Parse JSON content if possible, otherwise store as string
            try:
                parsed = json.loads(msg["content"])
                if isinstance(parsed, list):
                    tc["result"] = {"count": len(parsed), "sample": parsed[:3]}
                else:
                    tc["result"] = parsed
            except (json.JSONDecodeError, TypeError, ValueError):
                tc["result"] = msg["content"]
            break


def _fire_and_forget_emit(config, opts, messages, result, model=None):
    session = opts["session"]
    normalized = session.record(result)

    # Extract tool results from the messages array
    _extract_tool_results(session, messages)

    # Capture system prompt from messages (first system message, only once)
    if not session._system_prompt and messages:
        for msg in (messages if isinstance(messages, list) else []):
            if isinstance(msg, dict) and msg.get("role") == "system" and msg.get("content"):
                session._system_prompt = msg["content"]
                break

    # If Workers AI didn't include model in the response, use the one passed to run()
    if model and not normalized.get("model"):
        session._model = model

    # When auto_emit is disabled, the caller controls event emission.
    # The wrapper still tracks usage/tool calls via session.record() above.
    if opts.get("auto_emit") is False:
        return

    # Accumulate tool-call rounds without emitting — only emit when there's
    # actual text content (the final response in a multi-round tool loop).
    if not normalized.get("content") and normalized.get("tool_calls"):
        return

    raw_content = ""
    if isinstance(messages, list):
        user_msgs = [m for m in messages if isinstance(m, dict) and m.get("role") == "user"]
        if user_msgs:
            raw_content = user_msgs[-1].get("content", "")

    if isinstance(raw_content, list):
        user_msg = "\n".join(
            b.get("text", "") for b in raw_content if isinstance(b, dict) and b.get("type") == "text"
        )
    else:
        user_msg = raw_content

    assistant_msg = normalized.get("content") or ""

    def _do_emit():
        try:
            session.emit_chat_turn(
                user_message=user_msg,
                assistant_response=assistant_msg,
            )
        except Exception as e:
            print(f"[pentatonic-ai] emit failed: {e}", file=sys.stderr)

    wait_until = opts.get("wait_until")
    if callable(wait_until):
        # On runtimes that terminate early (e.g., serverless), wait_until
        # keeps the process alive for background work.
        wait_until(_do_emit)
    else:
        _do_emit()


class _WrappedOpenAICompletions:
    def __init__(self, config, completions, opts):
        self._config = config
        self._completions = completions
        self._opts = opts

    def create(self, **kwargs):
        result = self._completions.create(**kwargs)

        # URL rewriting on text content
        session = self._opts["session"]
        if isinstance(result, dict):
            choices = result.get("choices") or []
            if choices:
                msg = choices[0].get("message") or {}
                content = msg.get("content")
                if content:
                    msg["content"] = rewrite_urls(
                        content, self._config,
                        session.session_id, self._opts.get("metadata"),
                    )

        _fire_and_forget_emit(self._config, self._opts, kwargs.get("messages"), result)
        return result

    def __getattr__(self, name):
        return getattr(self._completions, name)


class _WrappedOpenAIChat:
    def __init__(self, config, chat, opts):
        self._config = config
        self._chat = chat
        self.completions = _WrappedOpenAICompletions(config, chat.completions, opts)

    def __getattr__(self, name):
        if name == "completions":
            return self.completions
        return getattr(self._chat, name)


class _OpenAIWrapper:
    def __init__(self, config, client, opts):
        self._config = config
        self._client = client
        self._opts = opts
        self.chat = _WrappedOpenAIChat(config, client.chat, opts)

    @property
    def session_id(self):
        return self._opts["session"].session_id

    @property
    def tes_session(self):
        return self._opts["session"]

    def __getattr__(self, name):
        if name == "chat":
            return self.chat
        return getattr(self._client, name)


class _WrappedAnthropicMessages:
    def __init__(self, config, messages, opts):
        self._config = config
        self._messages = messages
        self._opts = opts

    def create(self, **kwargs):
        result = self._messages.create(**kwargs)

        # URL rewriting on text content blocks
        session = self._opts["session"]
        if isinstance(result, dict) and isinstance(result.get("content"), list):
            for block in result["content"]:
                if isinstance(block, dict) and block.get("type") == "text" and block.get("text"):
                    block["text"] = rewrite_urls(
                        block["text"], self._config,
                        session.session_id, self._opts.get("metadata"),
                    )

        _fire_and_forget_emit(self._config, self._opts, kwargs.get("messages"), result)
        return result

    def __getattr__(self, name):
        return getattr(self._messages, name)


class _AnthropicWrapper:
    def __init__(self, config, client, opts):
        self._config = config
        self._client = client
        self._opts = opts
        self.messages = _WrappedAnthropicMessages(config, client.messages, opts)

    @property
    def session_id(self):
        return self._opts["session"].session_id

    @property
    def tes_session(self):
        return self._opts["session"]

    def __getattr__(self, name):
        if name == "messages":
            return self.messages
        return getattr(self._client, name)


class _WorkersAIWrapper:
    def __init__(self, config, ai_binding, opts):
        self._config = config
        self._ai = ai_binding
        self._opts = opts

    @property
    def session_id(self):
        return self._opts["session"].session_id

    @property
    def tes_session(self):
        return self._opts["session"]

    def run(self, model, params=None, **kwargs):
        result = self._ai.run(model, params, **kwargs)

        # URL rewriting on response text
        session = self._opts["session"]
        if isinstance(result, dict) and result.get("response"):
            result["response"] = rewrite_urls(
                result["response"], self._config,
                session.session_id, self._opts.get("metadata"),
            )

        messages = params.get("messages") if isinstance(params, dict) else None
        _fire_and_forget_emit(self._config, self._opts, messages, result, model=model)
        return result

    def __getattr__(self, name):
        return getattr(self._ai, name)
